projects/dqn/src/rl_experiment.py

```diff
@@ -0,0 +1,278 @@
+        algorithm_args = config.get("algorithm_args", {})
+        try:
+            num_actions = self.dataset.action_space.n
+        except ValueError:
+            self.logger.warning("This implementations expects a discrete action space")
+        algorithm_args["num_actions"] = num_actions
+        algorithm_args["observation_size"] = \
+            self.dataset.observation_space.shape
+
+        # Num steps required for algorithm schedulers
+        self.num_steps = config.get("num_steps", None)
+        if self.num_steps is None:
+            raise ValueError("Must define num_steps")
+        algorithm_args["num_steps"] = self.num_steps
+
+        # Define reward clipping, if existing
+        self.reward_clip = config.get("reward_clip", None)
+
+        # Initialize algorithm
+        algorithm_class = config.get("algorithm_class", None)
+        if algorithm_class is None:
+            raise ValueError("Must define algorithm_class")
+
+        self.algorithm = algorithm_class(
+            **algorithm_args,
+            device=self.device
+        )
+
+        self.replay_buffer = ReplayBuffer()
+        self.buffer_batch_size = config.get("buffer_batch_size", 64)
+        self.gradient_steps_per_epoch = config.get("gradient_steps_per_epoch", 16)
+
+        if self.dataset is None:
+            raise ValueError("Dataset must be initialized beforehand.")
+
+        self.data_collection_loader = self.create_data_collection_loader()
+
+        self.num_episodes_per_epoch = config.get("num_episodes_per_epoch", 1)
+
+    def collect_data(self, num_episodes):
+        """Return the mean of episode rewards"""
+
+        episode_rewards = Counter()
+        # how to run one step in data collection?
+        for episode in range(num_episodes):
+            for experience in self.data_collection_loader:
+                observation, action, reward, done, next_observation, info = experience
+                # Clip rewards and other transformations from the agent side
+                if self.reward_clip is not None:
+                    min_reward, max_reward = self.reward_clip
+                    reward = max(min_reward, min(reward, max_reward))
+                episode_rewards[episode] += reward
+                # Append to buffer
+                self.replay_buffer.append(experience)
+
+        return np.mean(list(episode_rewards.values()))
+
+    def pre_epoch(self):
+        pass
+
+    def post_epoch(self):
+        pass
+
+    def run_epoch(self):
+
+        self.pre_epoch()
+
+        # Collect Experiences
+        ret = {}
+        t0 = time.time()
+        ret["average_reward"] = self.collect_data(self.num_episodes_per_epoch)
+
+        # Train
+        t1 = time.time()
+        ret["average_loss"] = self.train_epoch()
+        t2 = time.time()
+
+        self.post_epoch()
+
+        self.logger.debug("data collection time: %s", t1 - t0)
+        self.logger.debug("training time: %s", t2 - t1)
```

Comment:
This timing would only work for CPU benchmarking. CUDA is asynchronous.
To measure GPU time you need to use `torch.cuda.Event` and `torch.cuda.synchronize`

Reply:
Even though it is asynchronous, I still have to wait until all is done to return the loss in `train_epoch`. Even if inside  `collect_data` and `train_epoch` there is parallelization, that should not change the time collected in the outer loop. Not sure I understand your point here.